### PR TITLE
Prevent double writes of samples to IPFS

### LINKED ISSUE #204: Write block data to IPFS asynchronously
Per a [todo](https://github.com/lazyledger/lazyledger-core/blob/417a817f62650951db713a1d6a6611c87bff4c5d/consensus/state.go#L1112) in #178 we should investigate the consequences of posting data asynchronously. This will likely be the case in the future, particularly when block sizes begin to increase.

### PR DESCRIPTION
**Pull request by Wondertan**
## Problem
As we compute the NMT tree for columns and rows of shares in an extended data square, we end up adding all the shares into IPFS two times, thus introducing unnecessary overhead, especially for IO. Considering how big blocks may become and how quickly they are generated, the overhead would scale significantly.

## Solution
The PR introduces a simple patch in NMTNodeAdder to memoize added leaves not to be added a second time.

EDIT:
It also fixes incorrect use of `BatchAdder`

EDIT2:
Closes #204

### CODE REVIEW

File: types/block.go
Comment (Wondertan) [inline types/block.go:324]: @evan-forbes, what was the reason to create a batch per `leafSet`? I was trying to find the answer myself, but I couldn't. Having one adder for all nodes seems like a better way to go for me.

Comment (evan-forbes) [reply types/block.go:324]: > what was the reason to create a batch per leafSet? 

each `leafSet` is all of the leaf data for a single row or column of the extended data square, so this is how we get each row and col root that go into the data availability header. While we will eventually write these batches asynchronously #204, this is needed using the current plugin/nmt structure.

Alternatively, I think we could write all of the data in a single batch using a different `NmtNodeAdder` structure, one that shares a batch for all each nmt.

Does that make any sense?

Comment (Wondertan) [reply types/block.go:324]: I think the description of the `format.Batch` is misleading. You don't need to create a new `Batch` every time you have some group of nodes to be added. Instead, the `Batch` separates multiple additions into batches itself. So once a single `Batch` is created, you can use it extensively to add a lot of nodes.

Comment (Wondertan) [reply types/block.go:324]: So regarding a TODO:

>// TODO: post data to IPFS in a goroutine

I don't think it is necessary as well, as `Add` and `AddMany` are almost never blocking for `Batch`.

Comment (Wondertan) [reply types/block.go:324]: Once optionate Batch buffer [limit exceeds](https://github.com/ipfs/go-ipld-format/blob/master/batch.go#L136), it runs [async node addition](https://github.com/ipfs/go-ipld-format/blob/master/batch.go#L99). Also, it [only blocks](https://github.com/ipfs/go-ipld-format/blob/master/batch.go#L87) when there are `NumCpu` in-progress parallel batches.

Comment (Wondertan) [reply types/block.go:324]: Does that make any sense?

Comment (evan-forbes) [reply types/block.go:324]: ahh I see now! You're right, that's awesome!

**Review by evan-forbes — APPROVED** [review]
I'd wait for another reviewer to comment before merging, but this LGTM :+1: 

IMO this closes #204 as well.

File: p2p/ipld/plugin/nodes/nodes.go
Comment (liamsi) [inline p2p/ipld/plugin/nodes/nodes.go:228]: This is cool! Wouldn't it make sense to check the same for leaves as well? If you look at the extended data square, e.g. the leaf at (0, 0) is contained in the first row and the first column. So with your changes, it seems it should make sense as there is only one batchAdder per extended block now?

File: p2p/ipld/plugin/nodes/nodes.go
Comment (liamsi) [inline p2p/ipld/plugin/nodes/nodes.go:210]: Optional nit; reorder while at it:
```suggestion
	ctx   context.Context
	batch *format.Batch
	leaves *cid.Set
```

File: p2p/ipld/plugin/nodes/nodes.go
Comment (Wondertan) [reply p2p/ipld/plugin/nodes/nodes.go:228]: > wouldn't it make sense to check the same for leaves as well (case 2 below)?

Yeah, it makes sense now, but only for leaf nodes, not non-leaf ones(case 2).

Comment (Wondertan) [reply p2p/ipld/plugin/nodes/nodes.go:228]: From my understanding we should not get any duplicates for case 2

Comment (Wondertan) [reply p2p/ipld/plugin/nodes/nodes.go:228]: Friendly ping @liamsi to understand if this is resolved

Comment (liamsi) [reply p2p/ipld/plugin/nodes/nodes.go:228]: You are absolutely right!

**Review by liamsi — APPROVED** [review]
Great find and great work! 🚀

**Pull request merged**
